### PR TITLE
gh-102153: fix CVE-2023-24329

### PR DESCRIPTION
**Pull request by xiaoge1001**
fix CVE-2023-24329

* Issue: gh-102153

### CODE REVIEW

**Comment by xiaoge1001** [comment]
```pycon
>>> from urllib.parse import urlparse, urlsplit
>>> urlparse(" https://example.com")
ParseResult(scheme='', netloc='', path=' https://example.com', params='', query='', fragment='')
```

```pycon
>>> from urllib.parse import urlparse, urlsplit
>>> urlparse(" https://example.com")
ParseResult(scheme='https', netloc='example.com', path='', params='', query='', fragment='')
```

File: Misc/NEWS.d/next/Security/2023-03-06-22-48-08.gh-issue-102153.eiaVrE.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+fix CVE-2023-24329
```
Comment (CharlieZhao95) [inline Misc/NEWS.d/next/Security/2023-03-06-22-48-08.gh-issue-102153.eiaVrE.rst]:
```suggestion
Fix the parsing problem(CVE-2023-24329) in urlparse when the entire URL
starts with blank characters. This vulnerability would help an attacker to bypass
the protections set by the developer for scheme and host.
```

**Review by CharlieZhao95 — CHANGES_REQUESTED** [review]

**Comment by CharlieZhao95** [comment]
BTW, this solution is mentioned at the end of the issue finder's article, see: https://pointernull.com/security/python-url-parse-problem.html for more details.

**Comment by xiaoge1001** [comment]
> BTW, this solution is mentioned at the end of the issue finder's article, see: https://pointernull.com/security/python-url-parse-problem.html for more details.

Yes, I read this article, too.  About the leading blank problem, I can solve it using lstrip() function even without reading this article.

File: Misc/NEWS.d/next/Security/2023-03-06-22-48-08.gh-issue-102153.eiaVrE.rst (hunk 1 above)
Comment (xiaoge1001) [reply Misc/NEWS.d/next/Security/2023-03-06-22-48-08.gh-issue-102153.eiaVrE.rst]: It's very great. I change it, thanks

**Comment by xiaoge1001** [comment]
https://github.com/python/cpython/issues/102153#issuecomment-1458022812

**Comment by illia-v** [comment]
> [#102153 (comment)](https://github.com/python/cpython/issues/102153#issuecomment-1458022812)
> 
> As described by @illia-v, I think it is necessary to remove the leading blanks from the URL in urllib.parse.

Right, but a few characters not covered by simple `.lstrip()` have to be stripped too, I created an alternative – #102508.

**Comment by alittlesir** [comment]
i think that the PR of @xiaoge1001 is a good idea,  could you review the PR, my brother @gpshead

**Comment by hugovk** [comment]
Thank you for the PRs, someone will review them and let's be patient in the meantime. 

See also https://github.com/python/cpython/issues/102153#issuecomment-1449143107

**Comment by xiaoge1001** [comment]
@gpshead  I'm sorry to bother you, but I don't know urllib very well. Can you answer my question? 

If I use this solution to fix CVE when we don't think about compatibility, can the CVE be completely fixed? Will other problems be introduced?

**Comment by gpshead** [comment]
We aren't currently confident that we can safely backport a fix to older Python versions as _something_ likely depends on some of the behavior around spaces.  That said, this solution likely does work for a lot of code.

It is a reasonable recommendation (as I think the blog post suggested?) for people to call .lstrip() themselves on their url before passing it to urlparse from their own code rather than assuming they're running on top of a Python without the issue.

The other PR I closed this in favor of will have potentially even more side effects that could impact other uses of urlparse as it removes a wider/different set of characters than `lstrip()` considers from anywhere in the entire url string.

If you do go with this or any of the PRs in your own patched Python interpreter, please let us know how that went. Details like the relative size of your overall Python code base using your patched interpreter was and how many, if any, issues around the behavior change in the API came up after some period of time testing and deploying things with it. Those are all good data to have to inform decisions. (I'll be trying some testing of that nature myself)

**Comment by illia-v** [comment]
> The other PR I closed this in favor of will have potentially even more side effects that could impact other uses of urlparse as it removes a wider/different set of characters than `lstrip()` considers from anywhere in the entire url string.

@gpshead yes, it may have more side effects. But it removes the characters only at the beginning and the end of a URL string.

**Pull request closed without merging**